### ISSUE TITLE
Version Assemblies task no longer does multiline matches (due to update to sed in shelljs)

### ISSUE THREAD

**Issue by ColinDabritzViewpoint** [colindembovsky/cols-agent-tasks#57]
**tldr: Version Assemblies had per-file matches previously, but now has per-line matches in the latest version. Workaround for some cases: collapse settings into single lines.**

Thanks much for the excellent and useful task library!

We've been using the versioning task for iOS mobile `Info.plist` versioning for some time now. The "Version Assemblies" task has been working for a while with version 2.0.86, but just recently stopped working in 2.2.117. We have tracked this down to a change in sed in shelljs upstream.

Issue Analysis
=========

By default, for iOS the `Info.plist` file has keys and values on separate lines, which is awkward in terms of XML structure, but worked fine. Our replacement is a little odd because of this.

We're replacing these lines:
```
	<key>CFBundleVersion</key>
	<string>2.16.2.0</string>
```

and using the settings as outlined in our readme contribution for iOS:
https://github.com/colindembovsky/cols-agent-tasks/blob/master/Tasks/VersionAssemblies/readme.md

Including the replacement pattern: `<key>CFBundleVersion</key>\s*<string>.*</string>`

The script log reports that it found one instance, and replaced it, and the displayed values look correct for regex, build number, and replacements.

```
2017-04-24T11:19:56.6690231Z ==============================================================================
2017-04-24T11:19:56.8721473Z Using prefix [<key>CFBundleVersion</key><string>] and version [2.16.2.64] and postfix [</string>] in folder [C:\agent\_work\14\s\FieldViewMobile.iOS]
2017-04-24T11:19:56.9815246Z Changing version in C:\agent\_work\14\s\FieldViewMobile.iOS\Info.plist
2017-04-24T11:19:56.9815246Z 1 matches for regex [<key>CFBundleVersion</key>\s*<string>.*</string>] found in file C:\agent\_work\14\s\FieldViewMobile.iOS\Info.plist
2017-04-24T11:19:56.9815246Z Processed 1 files
2017-04-24T11:19:56.9971450Z ##[section]Finishing: Version iOS info.plist (full version)
```

system.debug = True in variables doesn't appear to offer better information.

```
2017-04-24T17:06:10.4037584Z ##[debug]608 results
2017-04-24T17:06:10.4037584Z ##[debug]found 608 paths
2017-04-24T17:06:10.4037584Z ##[debug]applying include pattern
2017-04-24T17:06:10.4037584Z ##[debug]adjustedPattern: 'C:\agent\_work\14\s\FieldViewMobile.iOS\**\info.plist'
2017-04-24T17:06:10.4193833Z ##[debug]1 matches
2017-04-24T17:06:10.4193833Z ##[debug]1 final results
2017-04-24T17:06:10.4193833Z Changing version in C:\agent\_work\14\s\FieldViewMobile.iOS\Info.plist
2017-04-24T17:06:10.4193833Z 1 matches for regex [<key>CFBundleVersion</key>\s*<string>.*</string>] found in file C:\agent\_work\14\s\FieldViewMobile.iOS\Info.plist
2017-04-24T17:06:10.4193833Z Processed 1 files
2017-04-24T17:06:10.4193833Z ##[debug]Leaving Version Assemblies step
2017-04-24T17:06:10.4350096Z ##[section]Finishing: Version iOS info.plist (full version)
```

Checking the file on disk shows no replacement occurred.

Suspected cause
===========
I believe this was exposed when shelljs was updated from version 0.5.3 to 0.7.7 in this changeset:

https://github.com/colindembovsky/cols-agent-tasks/commit/14816af51f8ebd8a8098b899e55dacb1fc7f040c

This change in shelljs is in 0.7.7, but not in 0.5.3:
https://github.com/shelljs/shelljs/commit/1ee696d824e5f6bd9fd581a4f5f4910dd4e2d41d

and it fairly explicitly converts the regex replacement in sed to use split('\n'), and become a per-line editor (as sed generally is by default). The function has been changed further since, but this was when the 'whole file' behavior became 'per line'.

I did not see an option to change the behavior of sed in this case, so I am unsure how to proceed. For the versioning task, it might be helpful to add an option for multi-line matching vs single line matching.

**Comment by colindembovsky** [comment]
Thanks for reporting this @ColinDabritzViewpoint - I'll take a look at this. Including a multi-line option sounds like a good solution.

**Comment by colindembovsky** [comment]
Hi @ColinDabritzViewpoint I'm not sure how the `sh.sed` could ever have worked, but I could repro and couldn't get it to work. I eventually used `string.replace` instead of the `sh.sed` call and it seems to be working.

**Comment by ColinDabritzViewpoint** [comment]
See this change set in sh.sed:
https://github.com/shelljs/shelljs/commit/1ee696d824e5f6bd9fd581a4f5f4910dd4e2d41d

The previous implementation was:
`var result = fs.readFileSync(file, 'utf8').replace(regex, replacement);`

So string.replace is apparently how it used to work in 0.5.3. The newer version starts breaking out lines, and changes the behavior we depended on.

We're on VSTS, and it looks like 2.x has updated for us, we're seeing version 2.2.119

```
2017-04-26T16:37:36.4839826Z ==============================================================================
2017-04-26T16:37:36.4839826Z Task         : Version Assemblies
2017-04-26T16:37:36.4839826Z Description  : Updates the version number of the assemblies to match the build number
2017-04-26T16:37:36.4839826Z Version      : 2.2.119
2017-04-26T16:37:36.4839826Z Author       : Colin Dembovsky (colinsalmcorner.com)
2017-04-26T16:37:36.4839826Z Help         : [More Information](https://github.com/colindembovsky/cols-agent-tasks/tree/master/Tasks/VersionAssemblies)
2017-04-26T16:37:36.4839826Z ==============================================================================
```

I was able to run our existing work-around build as well as a new build which reverted the work around (and would have failed previously).

Both versions have gotten the correct replacements.

This preserves our existing behavior dependency in 2.x.

Thank you for your curation of this excellent set of tasks, and your quick responses, it is greatly appreciated!

I think we're good to go, and this issue is resolved.

**Comment by colindembovsky** [comment]
Thanks @ColinDabritzViewpoint - I noticed the line break-out in shelljs, and stole the `string.replace` idea from the previous version :-). Glad you're unstuck!
